hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java

```diff
@@ -107,21 +109,55 @@
 import static com.hazelcast.util.FutureUtil.logAllExceptions;
 import static com.hazelcast.util.IterableUtil.nullToEmpty;
 import static com.hazelcast.util.Preconditions.checkNotNull;
+import static java.lang.Math.ceil;
+import static java.lang.Math.log10;
 import static java.lang.Math.min;
+import static java.lang.Math.round;
 import static java.util.Collections.singleton;
 import static java.util.concurrent.TimeUnit.SECONDS;
 import static java.util.logging.Level.WARNING;
 
 abstract class MapProxySupport extends AbstractDistributedObject<MapService> implements InitializingObject {
 
     protected static final String NULL_KEY_IS_NOT_ALLOWED = "Null key is not allowed!";
-
     protected static final String NULL_VALUE_IS_NOT_ALLOWED = "Null value is not allowed!";
     protected static final String NULL_PREDICATE_IS_NOT_ALLOWED = "Predicate should not be null!";
     protected static final String NULL_LISTENER_IS_NOT_ALLOWED = "Null listener is not allowed!";
 
     private static final int CHECK_IF_LOADED_TIMEOUT_SECONDS = 60;
 
+    /**
+     * Defines the batch size for operations of {@link IMap#putAll(Map)} calls.
+     *
+     * A value of {@code 0} disables the batching and will send a single operation per member with all map entries.
+     *
+     * If you set this value too high, you may ran into OOME or blocked network pipelines due to huge operations.
+     * If you set this value too low, you will lower the performance of the putAll() operation.
+     */
+    private static final HazelcastProperty MAP_PUT_ALL_BATCH_SIZE
+            = new HazelcastProperty("hazelcast.map.put.all.batch.size", 0);
+
+    /**
+     * Defines the initial size of arrays per partition for {@link IMap#putAll(Map)} calls.
+     *
+     * The default value of {@code 0} uses an educated guess, depending on the map size, which is a good overall strategy.
+     * If you insert entries which don't match a normal partition distribution you should configure this factor.
+     * The initial size is calculated by this formula:
+     * {@code initialSize = ceil(MAP_PUT_ALL_INITIAL_SIZE_FACTOR * map.size() / PARTITION_COUNT)}
+     *
+     * As a rule of thumb you can try the following values:
+     * <ul>
+     * <li>{@code 10} for map sizes about 100 entries</li>
+     * <li>{@code 5} for map sizes between 500 and 5000 entries</li>
+     * <li>{@code 1.5} for map sizes between about 50000 entries</li>
+     * </ul>
+     *
+     * If you set this value too high, you will waste memory.
+     * If you set this value too low, you will suffer from expensive {@link java.util.Arrays#copyOf} calls.
+     */
+    private static final HazelcastProperty MAP_PUT_ALL_INITIAL_SIZE_FACTOR
```

Comment:
These properties are a bit in a hidden location. Normally user modifiable properties are part of GroupProperties (even though the class is internal). These will be included in the documentation; however your properties are not. This could be good; exposing more than strictly needs to be could lead to problems.